netbox/netbox/api/viewsets/__init__.py

```diff
@@ -41,8 +33,11 @@ def initial(self, request, *args, **kwargs):
 
         # Restrict the view's QuerySet to allow only the permitted objects
         if request.user.is_authenticated:
-            if action := HTTP_ACTIONS[request.method]:
-                self.queryset = self.queryset.restrict(request.user, action)
+            for permission in self.get_permissions():
+                if hasattr(permission, 'get_action') and (action := permission.get_action(request.method)):
+                    self.queryset = self.queryset.restrict(request.user, action)
+                elif action := HTTP_ACTIONS[request.method]:
+                    self.queryset = self.queryset.restrict(request.user, action)
```

Comment:
Can we just call `permission.get_action()`? Is there a scenario where we wouldn't be using TokenPermissions?

Reply:
We could, I was primarily worried there might be a case where there wouldn't be, but I was even on the fence about this.